### PR TITLE
os-release file is now used for hard classes and sys.flavor on all li…

### PR DESCRIPTION
**Pull request by olehermanse**
Edit: Read updated commit message in commit log.

### CODE REVIEW

**Comment by olehermanse** [comment]
@nickanderson Please review commit message.

Does it make more sense to set flavor to `coreos_1185`/`ubuntu_16`? If so, I can fix that.

@vpodzime Please review code.

**Comment by olehermanse** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/499/)

**Comment by olehermanse** [comment]
@nickanderson Fixed!

**Comment by olehermanse** [comment]
@nickanderson @craigcomstock `url_get.cf` failed again for a travis build on this PR.

File: libenv/sysinfo.c [hunk 1]
```diff
@@ -1058,6 +1058,51 @@ static void SetFlavour(EvalContext *ctx, const char *flavour)
...
+static void SetFlavour2(
+    EvalContext *const ctx,
+    const char *const id,
+    const char *const major_version)
+{
+    assert(ctx != NULL);
+    assert(id != NULL);
+    assert(major_version != NULL);
+
+    char *flavor;
```
Comment (vpodzime) [inline libenv/sysinfo.c]: Please stick to the *flavour* spelling everywhere.

File: libenv/sysinfo.c [hunk 2]
```diff
@@ -1058,6 +1058,51 @@ static void SetFlavour(EvalContext *ctx, const char *flavour)
...
+    xasprintf(&flavor, "%s_%s", id, major_version);
+    SetFlavour(ctx, flavor);
+    free(flavor);
+}
+
+// Combine version and id into one long canonified string,
+// then define n + 1 hard classes where n is the number of underscores
+// NOTE: strings should be canonified before calling
```
Comment (vpodzime) [inline libenv/sysinfo.c]: Why not use doxygen-style comments? (also applies to the other new function)

File: libenv/sysinfo.c [hunk 3]
```diff
@@ -1058,6 +1058,51 @@ static void SetFlavour(EvalContext *ctx, const char *flavour)
...
+    xasprintf(&class, "%s_%s", id, version);
+
+    // Strip away version number to define multiple hard classes
+    // Example: coreos_1185_3_0 -> coreos_1185_3 -> coreos_1185 -> coreos
+    char *last_underscore;
+    while ( (last_underscore = strrchr(class, '_')) != NULL )
+    {
+        EvalContextClassPutHard(ctx, class, tags);
+        *last_underscore = '\0';
+    }
```
Comment (vpodzime) [inline libenv/sysinfo.c]: I'd prefer this to be:
```
char * last_underscore = strrchr(class, '_');
while (last_underscore != NULL)
{
    ...
    *last_underscore = '\0';
    last_underscore = strrchr(class, '_');
}
```

File: libenv/sysinfo.c (hunk 1 above)
Comment (olehermanse) [reply libenv/sysinfo.c]: Yes it's a little annoying, but I would argue that we should default to american spelling. `sys.flavor` is much more used in our codebase, this file is pretty much the only culprit when it comes to using british spelling.

File: libenv/sysinfo.c (hunk 2 above)
Comment (olehermanse) [reply libenv/sysinfo.c]: I don't think this is a big deal for small static functions. If the comment was any longer than this (explaining the parameters etc.) I would use doxygen.

File: libenv/sysinfo.c (hunk 1 above)
Comment (olehermanse) [reply libenv/sysinfo.c]: I changed all occurences to flavor instead.

File: libenv/sysinfo.c (hunk 2 above)
Comment (olehermanse) [reply libenv/sysinfo.c]: Fixed, changed to doxygen.

File: libenv/sysinfo.c (hunk 3 above)
Comment (olehermanse) [reply libenv/sysinfo.c]: Fixed, thanks.

**Comment by nickanderson** [comment]
Looks good to me

**Comment by olehermanse** [comment]
Unrelated RHEL5 failure. @vpodzime is working on fixing / removing this platform for master branch. Merging.

**Pull request merged**